File: src/main/java/dev/alexengrig/leetcode/_350_intersection_of_two_arrays2/MapSolution.java
package dev.alexengrig.leetcode._350_intersection_of_two_arrays2;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

class MapSolution implements Task {
    @Override
    public int[] intersect(int[] nums1, int[] nums2) {
        int[] shortNums, nums;
        if (nums1.length <= nums2.length) {
            shortNums = nums1;
            nums = nums2;
        } else {
            shortNums = nums2;
            nums = nums1;
        }
        Map<Integer, Integer> counts = new HashMap<>();
        for (int num : shortNums) {
            counts.compute(num, (k, count) -> count == null ? 1 : count + 1);
        }
        int[] results = new int[shortNums.length];
        int index = 0;
        for (int i = 0, n = nums.length; !counts.isEmpty() && i < n; i++) {
            if (counts.containsKey(nums[i])) {
                counts.compute(nums[i], (k, count) -> count == null || count == 1 ? null : count - 1);
                results[index++] = nums[i];
            }
        }
        return Arrays.copyOf(results, index);
    }
}

File: src/main/python/350_intersection_of_two_arrays2/map_solution.py
from collections import Counter
from typing import List

from task import Task


class MapSolution(Task):
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        res = []
        is_lr = len(nums1) <= len(nums2)
        nums = nums2 if is_lr else nums1
        counts = Counter(nums1 if is_lr else nums2)
        for num in nums:
            if counts[num] > 0:
                res.append(num)
                counts[num] -= 1
        return res
